internal-client/src/main/java/org/apache/uniffle/client/factory/ShuffleServerClientFactory.java

```diff
@@ -75,9 +80,29 @@ public synchronized ShuffleServerClient getShuffleServerClient(
     return serverToClients.get(shuffleServerInfo);
   }
 
+  public synchronized void closeClients(Set<ShuffleServerInfo> candidates) {
```

Comment:
I think you misunderstand my thought. I hope the candidates are the deleted client candidates rather than being excluded clients.

```java
public synchronized void closeClients(Set<ShuffleServerInfo> candidates) {
clients.remove_all(candidates)
}
```

Reply:
We need use `synchronized` outside if we find the candidates outside of this method.  As follow
```java
ShuffleServerClientFactory shuffleServerClientFactory = ShuffleServerClientFactory.getInstance();
synchronized (shuffleServerClientFactory) {
  Set<ShuffleServerInfo> candidates = shuffleServerClientFactory.findClosableClients(getAllShuffleServers());
  shuffleServerClientFactory.closeClients(candidates);
}
```
It feels worse than the original implementation.

Reply:
From my thought, all clients are managed by the shuffleServerClientFactory, if in this class, we could maintain a structure to record the every client reference count num. if the client reference count is 0, then we could release it. 

If so, the +1 operation could be happened on the `getShuffleServerClient`. And -1 should be happened on the removing in shuffleWriteClientImpl.

How about this way?

Reply:
> If so, the +1 operation could be happened on the `getShuffleServerClient`. And -1 should be happened on the removing in shuffleWriteClientImpl.

I think this can't be achieved. To do this, we must invoke `getShuffleServerClient ` before we invoke the RPC interface every time  and -1 after each call to the RPC interface.